File: BackTracking/1182_부분수열의 합.cpp
#include <iostream>

using namespace std;

int result = 0;
int arr[20];

void DFS(int n, int s, int idx, int tot) {
	// 탈출 조건
	if (idx == n)
		return;

	// 타겟넘버가 되면
	if (tot + arr[idx] == s) 
		result++;

	DFS(n, s, idx + 1, tot + arr[idx]);
	DFS(n, s, idx + 1, tot);
	return;
}

int solution(int n, int s) {
	int answer = 0;

	DFS(n, s, 0, 0);
	answer = result;
	return answer;
}

int main(void) {
	int n, s;

	cin >> n >> s;
	for (int i = 0; i < n; ++i) 
		cin >> arr[i];
	
	int ret = solution(n, s);
	cout << ret;

	return 0;
}

File: BackTracking/1182_부분수열의 합.py
#입출력
from sys import stdin
#입력 
input = stdin.readline

n,s=map(int,input().split())
arr=list(map(int,input().split()))

#dfs 이용
def DFS(i,tot): 
    global cnt;
    # 인덱스가 범위 초과시
    if i >=n: 
        return
    
    #리스트의 원소 더하고
    tot+=arr[i]
    
    #목표와 같은지 비교
    if tot==s: 
        cnt+=1
    
    #그 전의 합과 인덱스+1로 탐색
    DFS(i+1,tot-arr[i]) 
    #합과 인덱스+1 로 탐색
    DFS(i+1,tot)

cnt=0;
DFS(0,0)
print(cnt)